### PR TITLE
Analyzer refactors, encapsulate `Relations`

### PR DESCRIPTION
**Pull request by Override-6**
This simple pull request refactors some structure names in order to be less misleading and more straightforward 

- `GlobalObjectId` => `RelationId` as they ARE relations identifier
- `SourceObjectId` => `SourceId`
- Created `LocalId` in order to pair with `RelationId`

- `Symbol::Global` => `Symbol::External` as "global" symbols are more of global variables, functions, defined in script's roots. This refactor better matches with the fact that a symbol can either be local to environment (thus `Symbol::Local`) or external (thus `Symbol::External`).

- `Object` => `Relation` as its identifier is `RelationId` that could be flustered with `ObjectId`.

- `Relations::objects` => `Relations::relations` to match `Object` => `Relation` refactor
  The relations' map were unexposed, and `Relations` now implements `Index<RelationId>` and `Index<RelationId>`

also refactored some `get_` methods to `find_` in `Variables` that weren't 0(1)

### CODE REVIEW

File: analyzer/src/environment/variables.rs [hunk 1]
```diff
@@ -31,52 +31,52 @@ pub struct Variables {
 
     /// Relations with external variables.
     /// The key is the variable Names, where value is the relation with another external environment's [Locals]
-    externals: HashMap<Name, GlobalObjectId>,
+    externals: HashMap<Name, RelationId>,
 }
 
 impl Variables {
-    /// Creates a new local variable.
+    /// Declares a new local variable.
```
Comment (syldium) [inline analyzer/src/environment/variables.rs:38]: This still creates a new local variable identifier.

File: analyzer/src/relations.rs [hunk 2]
```diff
@@ -46,85 +62,86 @@ pub struct ResolvedSymbol {
...
+            state: RelationState::Resolved(resolved),
         }
     }
 }
 
 /// A collection of objects that are tracked globally and may link to each other.
 #[derive(Debug, Default)]
 pub struct Relations {
-    /// The objects that need resolution that are tracked globally.
+    /// The tracked relations.
```
Comment (syldium) [inline analyzer/src/relations.rs:130]:
```suggestion
    /// The globally tracked relations.
```

**Review by syldium — CHANGES_REQUESTED** [review]

File: analyzer/src/relations.rs (hunk 2 above)
Comment (Override-6) [reply analyzer/src/relations.rs:130]: idk, there's no "local" relations so, maybe `tracked relations between environments` could be an option ?

Comment (syldium) [reply analyzer/src/relations.rs]: Local relations exist as they just share the same local id. *between environments* translates great the purpose of this system of relations.

File: vm/cmake-build-debug/CMakeFiles/clion-Debug-log.txt
Comment (syldium) [inline vm/cmake-build-debug/CMakeFiles/clion-Debug-log.txt:1]: Nope

**Review by syldium — CHANGES_REQUESTED** [review]

File: analyzer/src/relations.rs [hunk 3]
```diff
@@ -133,42 +150,64 @@ impl Relations {
...
-        self.objects
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (RelationId, &mut Relation)> {
+        self.relations
             .iter_mut()
             .enumerate()
-            .map(|(id, object)| (GlobalObjectId(id), object))
+            .map(|(id, relation)| (RelationId(id), relation))
+    }
+
+    /// Returns a mutable iterator over all the objects.
```
Comment (syldium) [inline analyzer/src/relations.rs:170]:
```suggestion
    /// Returns an immutable iterator over all the objects.
```

**Review by syldium — CHANGES_REQUESTED** [review]

**Review by syldium — APPROVED** [review]
There are two things that are difficult in computer science: naming and off-by-one errors.

**Pull request merged**